draft-ietf-quic-recovery.md

```diff
@@ -1368,7 +1377,15 @@ acked_packet from sent_packets.
        return
      if (congestion_window < ssthresh):
        // Slow start.
-       congestion_window += acked_packet.size
+       if (pacing):
+         congestion_window += acked_packet.size
+       else:
+         // If pacing is not used, limit the increase to
+         // 2*max_datagram_size per ACK as described in
+         // {{?RFC3465}}
+         if (cwnd_increase < 2 * max_datagram_size):
+           cwnd_increase += bytes_acked
+           congestion_window += bytes_acked
```

Comment:
Same issue as L1387 (`bytes_acked`)

Reply:
I think this may allow increasing cwnd close to (3 x max_datagram_size).

e.g. max_datagram_size = 1000
ack 1 = 999 bytes
ack 2 = 999 bytes
ack 3 = 999 bytes

If we have 3 acks each acks 999 bytes, this code result in 999x3 bytes cwnd increase. 

To calculate remaining bytes correctly, I would propose the following:

```
  OnAckReceivedCC():
     cwnd_incr_max = 2 * max_datagram_size
...
  OnPacketAckedCC(acked_packet):
...
         // If pacing is not used, limit the increase to
         // 2*max_datagram_size per ACK as described in
         // {{?RFC3465}}
         cwnd_incr = cwnd_incr_max - max(cwnd_incr_max - acked_packet.size, 0)
         congestion_window += cwnd_incr
         cwnd_incr_max -= cnwd_incr
```

Reply:
good point. How about the below:
```
 if (pacing):
         congestion_window += acked_packet.size
       else:
         // If pacing is not used, limit the increase to
         // 2*max_datagram_size per ACK as described in
         // {{?RFC3465}}
         if (cwnd_increase + acked_packet.size <=
             2 * max_datagram_size):
           cwnd_increase += acked_packet.size
           congestion_window += acked_packet.size
```

Reply:
There is a problem with my suggested code that the cwnd could end up being less than 2*MSS.

This is my updated proposal:
```
if (pacing):
         congestion_window += acked_packet.size
       else:
         // If pacing is not used, limit the increase to
         // 2*max_datagram_size per ACK as described in
         // {{?RFC3465}}
         if (cwnd_increase  < 2 * max_datagram_size):
           cwnd_increase = min(cwnd_increase + acked_packet.size,
                                            2 * max_datagram_size)
           congestion_window = cwnd_increase
```

Reply:
> There is a problem with my suggested code that the cwnd could end up being less than 2*MSS.
> 
> This is my updated proposal:
> 
> ```
> if (pacing):
>          congestion_window += acked_packet.size
>        else:
>          // If pacing is not used, limit the increase to
>          // 2*max_datagram_size per ACK as described in
>          // {{?RFC3465}}
>          if (cwnd_increase  < 2 * max_datagram_size):
>            cwnd_increase = min(cwnd_increase + acked_packet.size,
>                                             2 * max_datagram_size)
>            congestion_window = cwnd_increase
> ```

I think it will reset congestion_window to a small value (<=2 mss).

Reply:
@junhochoi could you give an example where it would be less than 2MSS?

Reply:
I mean this part:

> congestion_window = cwnd_increase

If you run my example above (3 acks 999 bytes each), cwin will be 2 x mss at the end with algorithms above.

I think you mean `+=`, not `=`, but if so it makes another issue (cwin will increase much faster than expected)

Reply:
Last line is problematic because it will reset cwnd to cwnd_increase which is capped to 2 x mss.
```congestion_window = cwnd_increase```

Reply:
Isn't capping to 2*MSS the desired congestion window? `=` is my intended proposal to cap the congestion window to 2MSS.

Reply:
I believe the basic intention of RFC3465 (section 2.3) is limiting cwnd *increase* during slow start per ACK,  (e.g. one ACK acknowledges 4 MSS --> only increase cwnd by 2 MSS, not 4 MSS), not resetting cwnd. If cwnd is reset to 2xMSS, it is basically same to idle timeout scenario.

Reply:
My bad. I would have to fix this.

Reply:
The below should work in all cases:
```
if (pacing):
         congestion_window += acked_packet.size
       else:
         // If pacing is not used, limit the increase to
         // 2*max_datagram_size per ACK as described in
         // {{?RFC3465}}
         if (cwnd_increase  < 2 * max_datagram_size):
           old_cwnd_increase = cwnd_increase
           cwnd_increase = min(cwnd_increase + acked_packet.size,
                                            2 * max_datagram_size)
           congestion_window += cwnd_increase - old_cwnd_increase
```

Reply:
See mine at https://github.com/quicwg/base-drafts/pull/3232#discussion_r345506253
Basically same approach as yours. Note that I removed `if` and use `max()` instead of `min()`. I think either way works fine.

Reply:
I pushed my proposed changes before reading your comment. :-) I am fine with both approaches.
This issue will be discussed at IETF either today or tomorrow; we can decide it there.